adafruit_bitmapsaver.py

```diff
@@ -54,44 +61,50 @@ def _write_dib_header(output_file, width, height):
         output_file.write(b"\x00")
 
 
-def _bytes_per_row(source_width):
+def _bytes_per_row(source_width: int) -> int:
     pixel_bytes = 3 * source_width
     padding_bytes = (4 - (pixel_bytes % 4)) % 4
     return pixel_bytes + padding_bytes
 
 
-def _rotated_height_and_width(pixel_source):
+def _rotated_height_and_width(pixel_source: Union[Bitmap, Display]) -> Tuple[int, int]:
     # flip axis if the display is rotated
     if isinstance(pixel_source, Display) and (pixel_source.rotation % 180 != 0):
-        return (pixel_source.height, pixel_source.width)
-    return (pixel_source.width, pixel_source.height)
+        return pixel_source.height, pixel_source.width
+    return pixel_source.width, pixel_source.height
 
 
-def _rgb565_to_bgr_tuple(color):
-    blue = (color << 3) & 0x00F8  # extract each of the RGB tripple into it's own byte
+def _rgb565_to_bgr_tuple(color: int) -> Tuple[int, int, int]:
+    blue = (color << 3) & 0x00F8  # extract each of the RGB triple into it's own byte
     green = (color >> 3) & 0x00FC
     red = (color >> 8) & 0x00F8
-    return (blue, green, red)
+    return blue, green, red
 
 
 # pylint:disable=too-many-locals
-def _write_pixels(output_file, pixel_source, palette):
+def _write_pixels(
+    output_file: BufferedWriter,
+    pixel_source: Union[Bitmap, Display],
+    palette: Optional[Palette],
```

Comment:
Is palette really optional here? it doesn't default to `None` and it looks like there is an attempt to access a value from it using square brackets inside of this function without any further check for None that I'm seeing.

Reply:
Palette is optional. The function takes either a (bitmap & palette) and follows a major branch or (Display & None) and follows a different major branch

Related: I added and "# handled by save_pixel's guardians" after I figured this out